### ISSUE TITLE
WsClient::on_disconnect is not cancel-safe/footgun

### ISSUE THREAD

**Issue by xlc** [paritytech/jsonrpsee#996]
https://github.com/paritytech/jsonrpsee/blob/a330dae1e377807e021bde237996eb3cb4159786/core/src/client/async_client/mod.rs#L260-L272

`on_disconnect` can only be called once which makes this code incorrect https://github.com/AcalaNetwork/subway/blob/1026a5110837f96564732922ada96c9800dd495d/src/client/mod.rs#L114

We need to either improve the docs & API design to ensure the repeated call either works as expected or fail with some error indicate repeated call is not supported.

**Comment by niklasad1** [comment]
Ok, I see now [tokio::sync::Mutex::lock](https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#method.lock) is not cancel-safe and indeed the docs are wrong regarding the cancel safety.

So for now you need use [futures::future::select](https://docs.rs/futures/latest/futures/future/fn.select.html) and handle the future which isn't resolved and poll it again in your look to workaround the cancel-safety issue when using `tokio::select!`.

Let us take a look whether we can replace this with something cancel-safe.

However, this still works it's just a matter of cancel-safety:

```rust
#[tokio::main]
async fn main() -> anyhow::Result<()> {
	tracing_subscriber::FmtSubscriber::builder().finish().try_init()?;

	let addr = run_server().await?;
	let url = format!("ws://{}", addr);

	let client = WsClientBuilder::default().build(&url).await?;
	let response: String = client.request("say_hello", rpc_params![]).await?;
	tracing::info!("response: {:?}", response);

	for _ in 0..10 {
		client.on_disconnect().await;
		tracing::info!("disconnected");
	}

	Ok(())
}

async fn run_server() -> anyhow::Result<SocketAddr> {
	let server = ServerBuilder::default().build("127.0.0.1:0").await?;
	let mut module = RpcModule::new(());
	module.register_method("say_hello", |_, _| Ok("lo"))?;
	let addr = server.local_addr()?;
	let handle = server.start(module)?;

	tokio::spawn(async move {
		tokio::time::sleep(std::time::Duration::from_secs(5)).await;
		handle.stop().unwrap();
	});

	Ok(addr)
}
```

**Comment by niklasad1** [comment]
Closed by #999
